refactor(language_server): move tools definitions to `Backend`

Sysix opened the pull request:
> This PR refactors the language server architecture by moving tool definitions (linter and formatter) from individual workers to the Backend. The ToolBuilder trait is redesigned to use a method that returns boxed trait objects, and builder structs are simplified to stateless types.

Now it should be easy to implement something like `Backend::new(vec![ServerLinterBuilder]);`

camc314 commented:
### Merge activity

* **Nov 12, 9:23 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Sysix committed:
refactor(language_server): move tools definitions to `Backend` (#15563)

> This PR refactors the language server architecture by moving tool definitions (linter and formatter) from individual workers to the Backend. The ToolBuilder trait is redesigned to use a method that returns boxed trait objects, and builder structs are simplified to stateless types.

Now it should be easy to implement something like `Backend::new(vec![ServerLinterBuilder]);`
